flips/20220817-add-fields-to-existing-definitions.md

```diff
@@ -0,0 +1,80 @@
+|---------------|:--------------------------------------------------------- |
+| **FLIP #**    | [NNN](https://github.com/onflow/flow/pull/NNN)            |
+| **Author(s)** | Deniz Mert Edincik (deniz@edincik.com)                    |
+|               | Austin Kline (austin@flowty.io)                           | 
+| **Sponsor**   |                                                           |
+| **Updated**   | 2022-08-17                                                |
+
+## Objective
+This proposed change will allow existing structs and resources to add new fields to them by
+using optional and default assignment embedded in the object definition itself.
+
+## Motivation
+
+One major challenge in smart contract design in cadence is the inability to add new fields to already deployed 
+structs and resources. Not having the ability to do this means that contracts either get extended 
+with data that exists in another contract, entirely new contracts are made that are essentially reprints of the
+original, except with the newly added fields, or some other workaround such as storing fields as dictionaries
+and building them into at runtime.
+
+Each of these workarounds come with their downsides. 
+- New contracts lead to complex migrations for applications and sunsetting existing contracts. 
+- Hosting new data in another contract leads to harder-to-follow code and added complexity. 
+- Factory patterns increase compute since objects must be built at runtime and also lose the benefits of type-checking
+    since the underlying structure is not truly typed.
+
+## User Benefit
+
+- It will allow developers to add fields as they need them, making contract development more focused on the needs of the current version
+    as opposed to undue complexity to take future plans into account.
+- It will allow more maintainable contract code with less splintered logic as existing contracts update themselves
+
+## Design Proposal
+
+We propose to add the ability to define new optional fields, and the assignment of default values to fields in
+structs and resources. All fields added after the first deployment of a contract must be either optional or have
+a default parameter assigned so that initialization of existing instances of these objects can be properly
+initialized.
+
+```cadence
+pub struct Message {
+    pub let content: String
+}
+```
+
+We could now add a new field to this struct in a few ways
+
+```cadence
+pub struct Message {
+    pub let content: String
+    
+    // new fields
+    // an optional new field. Existing instances of Message will have timestamp set to nil when accessed
+    // unless they are set at a later date
+    pub let timestamp: UInt64?
+    // a default initialized field, Existing instances of Message will take the default value.
+    pub let received: Bool = false 
+}
+```
+
+### Limitations
+
+- This will not allow existing fields to be altered. That is, you cannot take a field and alter its type.
```

Comment:
Users already have the ability to remove fields from existing composites, so giving them the ability to add new fields can implicitly allow changing a field's type if a user removes an existing field (that had type `T1`) in one update and then adds it back with a different type `T2` in another update. We'd need a specific solution for this case. One obvious answer is to simply fail a contract update if existing instances of a struct or resource have conflicting types with the field being added, but there are other potential solutions. Either way we should make it explicit in the proposal.

Reply:
We could probably propose similar functionality as proposed for type removal in https://github.com/onflow/flips/pull/276: If we propose to allow adding new fields, the removal of fields should be required to "tombstone" a field, so it cannot be re-added. 

For example, this could be done through something like a `#removedField(fieldName)` pragma.

Reply:
I like the idea of tomb-stoning. However there is a chance that people have already removed fields from composites without the pragma, and the data for those fields still exist. So adding a new field could still cause type-safety issues. Maybe we could cleanup the composite values as part of the migration, to remove fields that are not part of their corresponding type? (IDK if this is possible / how to do that, though)